server/service/src/sync/integrate_document.rs

```diff
@@ -0,0 +1,87 @@
+use repository::{
+    Document, DocumentRegistryFilter, DocumentRegistryRepository, DocumentRepository, EqualFilter,
+    RepositoryError, StorageConnection,
+};
+
+use crate::programs::{
+    encounter::{
+        encounter_updated::update_encounter_row, validate_misc::validate_encounter_schema,
+    },
+    patient::{patient_schema::SchemaPatient, patient_updated::patient_document_updated},
+    program_enrolment::program_enrolment_updated::program_enrolment_updated,
+    program_enrolment::program_schema::SchemaProgramEnrolment,
+};
+
+pub(crate) fn upsert_document(
+    con: &StorageConnection,
+    document: &Document,
+) -> Result<(), RepositoryError> {
+    let registry = DocumentRegistryRepository::new(con)
+        .query_by_filter(
+            DocumentRegistryFilter::new().document_type(EqualFilter::equal_to(&document.r#type)),
+        )?
+        .pop()
+        .ok_or(RepositoryError::NotFound)?;
+
+    DocumentRepository::new(con).insert(document)?;
+
+    match registry.context {
+        repository::DocumentContext::Patient => update_patient(con, document)?,
+        repository::DocumentContext::Program => update_program_enrolment(con, document)?,
+        repository::DocumentContext::Encounter => update_encounter(con, document)?,
+        repository::DocumentContext::Custom => {}
+    };
+    Ok(())
+}
+
+fn update_patient(con: &StorageConnection, document: &Document) -> Result<(), RepositoryError> {
+    let patient: SchemaPatient = serde_json::from_value(document.data.clone()).map_err(|err| {
+        RepositoryError::as_db_error(&format!("Invalid patient data: {}", err), "")
+    })?;
+
+    patient_document_updated(con, &document.datetime, patient)
+        .map_err(|err| RepositoryError::as_db_error(&format!("{:?}", err), ""))?;
+    Ok(())
+}
+
+fn update_program_enrolment(
+    con: &StorageConnection,
+    document: &Document,
+) -> Result<(), RepositoryError> {
+    let Some(patient_id) = &document.owner_name_id else {
+        return Err(RepositoryError::as_db_error("Document owner id expected", ""));
+    };
+    let program_enrolment: SchemaProgramEnrolment = serde_json::from_value(document.data.clone())
+        .map_err(|err| {
+        RepositoryError::as_db_error(&format!("Invalid program enrolment data: {}", err), "")
+    })?;
+    program_enrolment_updated(con, patient_id, document, program_enrolment)
+        .map_err(|err| RepositoryError::as_db_error(&format!("{:?}", err), ""))?;
+    Ok(())
+}
+
+fn update_encounter(con: &StorageConnection, document: &Document) -> Result<(), RepositoryError> {
+    let (Some(patient_id), Some(program)) = (&document.owner_name_id, &document.context) else {
+        return Err(RepositoryError::as_db_error("Document owner id and context expected", ""));
+    };
+
+    let encounter = validate_encounter_schema(&document.data).map_err(|err| {
```

Comment:
in validate_encounter_schema, there is this comment:

https://github.com/openmsupply/open-msupply/blob/3649f446b69758dfd4fa39d88fa7ef33f14ceeb2/server/service/src/programs/encounter/validate_misc.rs#L27-L28

Encounter row seems to always have one defined of schema, I am quite confused does it mean we are using document as another layer, kind of like domain object but generic/dynamic domain object used to translate to typed data types, and this layer is again used in sync. So instead of syncing enrolments and encounters we sync documents and then translate them to enrolments and encounters ?

Reply:
The document is kind of the source of truth and yes for `patients`, `encounters`, and `program_enrolments` they must follow a predefined base schema.

This guaranteed base schema allows us to fill `encounter` and `program_enrolment` tables. These tables are basically just aux tables. For sync we only need to sync the documents and then fill the encounter and enrolment table from the received documents.

> Just had a brief thought, the dashboard on central, they would need to look at json data to figure out details about enrolment/encounter ?

yes this is correct, the queries would have to work on json data

Reply:
> The document is kind of the source of truth and yes for patients, encounters, and program_enrolments they must follow a predefined base schema.

Ok, it makes sense, similar to the comment above, with confusion about syncing via document table, my concern was really about:

* With graphql we have a way to enforce schema (with inputs)
* With sync we have a way to validate and se/de to a concrete data type

Of course the above doesn't work for schema that is not know at runtime, but for schema that is know at runtime it has it's benefits. And I can see how documents can be used to extend a concrete data type, but what is the benefit of keeping the known schema (known at runtime), in the document, is it because it could be edited concurrently on two different sites ? Or because it's a generic way to store/sync it ?

Reply:
yes the editing and generic sync is one advantage but you also have an easy way to store the history and you are also able to easily ensure document integrity (calculating the hash). Furthermore, you can easily add other document types without a known schema.